containers/daap-landing-to-raw/src/var/task/validation.py

```diff
@@ -0,0 +1,86 @@
+class DataInvalid(Exception):
+    pass
+
+
+def type_is_compatible(registered_type: str, inferred_type: str) -> bool:
```

Comment:
Think this function needs some work still and more complex logic. e.g. if a `tinyint` is registered but a `bigint` is inferred then this should fail as `bigint` will be outside the range of `tinyint`. So `tinyint` registered should only pass for `tinyint` inferred, `smallint` for `tinyint` and `smallint`, etc

Reply:
Yeah, I'm aware this will let through some cases that aren't strictly valid, and I considered this for some time before I arrived at the current version.

It may be that there is some more cases we can reject, but it seems like too big of a big rabbit hole to go down at this stage of development.

In the scenario you mentioned, it may be that 90% of the data is a valid tinyint, but just 1 row has a bigint in, causing us to infer it as a bigint. In this situation, I think ultimately in this scenario we would want to accept the dataset, and treat it as a data quality issue later in the pipeline. Data quality is not in scope for the alpha, but still we made the decision to prefer lenient validation rather than strict validation, so I decided to keep the tinyint/smallint/int/bigint relatively simple.

It's also worthing keeping in mind that rAPId's implementation is similar but (at least for now) only has one special case (for dates iirc). So I think we've already gone quite far in trying to predict possible issues, and it shouldn't be too difficult to add in more special cases later once we get it in front of some users 🙂

Reply:
Hmm, but i'm not sure we could run data quality tests if the data are stored as in a scenario above because you won't be able to query that data at all. `select * from db.tbl` would just return an error, or any query with an impacted column. Could we not do something like assign a number to int types and then have some relatively simple if >= logic? so if tinyint were 1 and bigint 4 your case could have an `if registered_type => inferred_type` and return true if satisfied or else false 

Kind of thinking out loud. But if you think this wouldn't be appropriate i suppose for alpha we're unlikely to run into anything like this and it could always be developed further in beta

Reply:
Hi @LavMatt, I gave this a try this morning, but with the current schema inference code, there is no way to make it work.

This is because infer schema doesn't return the smallest possible integer for the dataset, it just always picks a bigint. This is down to the behaviour of arrow, for example:

```python
>>> csv=b"a,b,c\n1,2,3\n4,5,6"
>>> from mojap_metadata.converters.arrow_converter import ArrowConverter
from mojap_metadata.converters.glue_converter import GlueConverter
from mojap_metadata.metadata.metadata import Metadata
from pyarrow import csv as pa_csv
from pyarrow import parquet as pq>>> from mojap_metadata.converters.glue_converter import GlueConverter
>>> from mojap_metadata.metadata.metadata import Metadata
>>> from pyarrow import csv as pa_csv
>>> from pyarrow import parquet as pq
>>> from io import BytesIO
>>> f = BytesIO(csv)
>>> table = pa_csv.read_csv(f, convert_options=pa_csv.ConvertOptions(null_values=[""]))
>>> table
pyarrow.Table
a: int64
b: int64
c: int64
----
a: [[1,4]]
b: [[2,5]]
c: [[3,6]]
```

So arrow, will infer *A schema* but not *the strictest schema*. This means if the registered schema is `tinyint` we **cannot** know whether the data is going to overflow just by using infer schema on the data set.

For this reason I don't think we can make the validation any stricter without changing the infer schema logic.

For future reference this is the variant I tested

```python
INTEGER_WIDENESS = {"tinyint": 0, "smallint": 1, "int": 2, "integer": 2, "bigint": 3}

def type_is_compatible(registered_type: str, inferred_type: str) -> bool:
    """
    Validate a type inferred from the dataset is compatible with the schema.

    This implementation adds some leniency for types that are different, but probably compatible.
    For example: a `long` type is wider than an `integer` which is wider than `short` and `byte`.
    It is always valid for the dataset to contain a narrower type than what is registered in the schema,
    such as when the schema requires long but we detected integers.

    For a full list of supported types, see the AWS athena documentation:
    https://docs.aws.amazon.com/athena/latest/ug/data-types.html
    """
    if registered_type == inferred_type:
        return True

    match (registered_type, inferred_type):
        case ("string", _):
            # Anything can go in a string
            return True
        case ("timestamp", "date"):
            # Data can go in timestamp
            return True
        case (
            "tinyint" | "smallint" | "int" | "integer" | "bigint",
            "tinyint" | "smallint" | "int" | "integer" | "bigint",
        ) if INTEGER_WIDENESS[inferred_type] <= INTEGER_WIDENESS[registered_type]:
            # Integers can go in wider integers
            return True
        case (
            "float" | "double" | "decimal",
            "tinyint" | "smallint" | "int" | "integer" | "bigint",
        ):
            # Integers can go in floats
            return True
        case "double", "float":
            # Floats can go in doubles
            return True
        case _:
            return False
```